File: frontend/src/types/execution.ts
import { z } from "zod";

// Mirrors cron_dok.domain.entities.execution literals
export const executionStatusSchema = z.enum([
  "queued",
  "running",
  "succeeded",
  "failed",
  "skipped",
  "killed",
]);
export type ExecutionStatus = z.infer<typeof executionStatusSchema>;

export const triggerTypeSchema = z.enum(["scheduled", "manual"]);
export type TriggerType = z.infer<typeof triggerTypeSchema>;

// Mirrors ExecutionResponse (logs are served separately, spec 6.4)
export const executionSchema = z.object({
  id: z.number().int(),
  runner_id: z.number().int(),
  status: executionStatusSchema,
  trigger_type: triggerTypeSchema,
  started_at: z.string().nullable(),
  finished_at: z.string().nullable(),
  exit_code: z.number().int().nullable(),
  duration_ms: z.number().int().nullable(),
  log_path: z.string().nullable(),
});
export type Execution = z.infer<typeof executionSchema>;

// Mirrors LogChunkResponse (incremental log read)
export const logChunkSchema = z.object({
  chunk: z.string(),
  offset: z.number().int(),
});
export type LogChunk = z.infer<typeof logChunkSchema>;

File: backend/src/cron_dok/domain/entities/execution.py
"""Execution entity: metadata of one concrete run of a runner.

Logs live outside the database (spec section 6.4); only ``log_path`` is
stored here.
"""

from dataclasses import dataclass
from datetime import datetime
from typing import Literal

ExecutionStatus = Literal["queued", "running", "succeeded", "failed", "skipped", "killed"]
TriggerType = Literal["scheduled", "manual"]


@dataclass(kw_only=True)
class Execution:
    """A single run of a runner, scheduled or manual."""

    id: int | None = None
    runner_id: int
    status: ExecutionStatus = "queued"
    trigger_type: TriggerType = "scheduled"
    started_at: datetime | None = None
    finished_at: datetime | None = None
    exit_code: int | None = None
    duration_ms: int | None = None
    log_path: str | None = None
